src/main/java/seedu/mark/logic/commands/ImportCommand.java

```diff
@@ -0,0 +1,153 @@
+    public static String toIndentedString(List<Bookmark> bookmarks) {
+        String newlineAndIndent = "\n    ";
+        return bookmarks.stream().map(Bookmark::toString)
+                .map(bookmark -> newlineAndIndent.concat(bookmark))
+                .reduce("", String::concat);
+    }
+
+    /**
+     * Executes the import command on the given {@code model} and {@code storage}.
+     *
+     * @param model {@code Model} which the command should operate on.
+     * @param storage {@code Storage} which the command should operate on.
+     * @return A CommandResult indicating a successful import, and (if applicable)
+     *         shows bookmarks that were skipped during import.
+     * @throws CommandException if an error occurs while reading bookmarks from Storage
+     */
+    @Override
+    public CommandResult execute(Model model, Storage storage) throws CommandException {
+        requireAllNonNull(model, storage);
+
+        ReadOnlyMark newMark = readMarkFromStorageFile(storage, filePath);
+
+        ObservableList<Bookmark> bookmarksToImport = newMark.getBookmarkList();
+        List<Bookmark> skippedBookmarks = new ArrayList<>(); // empty list to operate on
+        addBookmarksToModel(model, bookmarksToImport, skippedBookmarks);
+
+        if (!skippedBookmarks.isEmpty()) {
+            logger.info("Bookmarks imported from " + filePath + ": Some duplicates skipped");
+            return new CommandResult(String.format(MESSAGE_IMPORT_SUCCESS_WITH_DUPLICATES, filePath,
+                    toIndentedString(skippedBookmarks)));
+        }
+
+        logger.info("Bookmarks imported from " + filePath);
+        return new CommandResult(String.format(MESSAGE_IMPORT_SUCCESS, filePath));
+    }
+
+    /**
+     * Attempts to read Mark from a given file in storage.
+     */
+    private ReadOnlyMark readMarkFromStorageFile(Storage storage, Path filePath) throws CommandException {
+        requireAllNonNull(storage, filePath);
+
+        Optional<ReadOnlyMark> newMark;
+        try {
+            newMark = storage.readMark(filePath);
+        } catch (IOException ioe) {
+            logger.info("Bookmarks not imported: Problem while reading from the file " + filePath);
+            throw new CommandException(MESSAGE_IMPORT_FAILURE);
+        } catch (DataConversionException dce) {
+            logger.info("Bookmarks not imported: Data file " + filePath + " has wrong format");
+            throw new CommandException(String.format(MESSAGE_FILE_FORMAT_INCORRECT, filePath));
+        }
+
+        assert newMark != null;
+        if (newMark.isEmpty()) {
+            throw new CommandException(String.format(MESSAGE_FILE_NOT_FOUND, filePath));
+        }
+
+        return newMark.get();
+    }
+
+    /**
+     * Adds bookmarks from the given list of bookmarks to the model and adds
+     * duplicate bookmarks to the list of skipped bookmarks.
+     */
+    private void addBookmarksToModel(Model model, ObservableList<Bookmark> bookmarksToAdd,
```

Comment:
I'm starting to have a feeling all these methods should be somewhere in the model, doesn't feel right to have so much logic going on here.

Reply:
Right... I'll go figure out where to put it. And likewise for `readMarkFromStorageFile` (moving to Storage).